Tensile/Source/lib/include/Tensile/Utils.hpp

```diff
@@ -37,6 +37,59 @@
 namespace Tensile
 {
 
+    // Type wrapper that can be copied or assigned to in a threadsafe manner. Value cannot be modified
+    template<typename T>
+    struct ThreadSafeValue
+    {
+    private:
+        mutable std::mutex access;
+        T                  value;
+    public:
+        ThreadSafeValue()
+        {
+
+        }
+
+        ThreadSafeValue(const ThreadSafeValue<T>& other)
+        {
+            std::lock_guard<std::mutex> lock (other.access);
+            value = other.value;
+        }
+
+        ThreadSafeValue(const T& other): value(other)
+        {
+
+        }
+
+        ThreadSafeValue<T>& operator=(const ThreadSafeValue<T>& other)
+        {
+            std::lock_guard<std::mutex> otherLock (other.access);
+            std::lock_guard<std::mutex> selfLock (access);
```

Comment:
You need to guard against self-assignment. This will deadlock if `this == &other`.